test/test_binary_ufuncs_ops.py

```diff
@@ -0,0 +1,258 @@
+            self.assertEqual(actual.item(), expected, **kwargs)
+        elif isinstance(expected, np.ndarray):
+            # Handles exact dtype comparisons between arrays and tensors
+            if exact_dtype:
+                # Allows array dtype to be float32 when comparing with bfloat16 tensors
+                #   since NumPy doesn't support the bfloat16 dtype
+                # Also ops like scipy.special.erf, scipy.special.erfc, etc, promote float16
+                # to float32
+                if expected.dtype == np.float32:
+                    assert actual.dtype in (torch.float16, torch.bfloat16, torch.float32)
+                else:
+                    assert expected.dtype == torch_to_numpy_dtype_dict[actual.dtype]
+
+            self.assertEqual(actual,
+                             torch.from_numpy(expected).to(actual.dtype),
+                             msg,
+                             exact_device=False,
+                             **kwargs)
+        else:
+            self.assertEqual(actual, expected, msg, exact_device=False, **kwargs)
+
+    # Tests that the function and its (array-accepting) reference produce the same
+    # values on given tensors
+    def _test_reference_numerics(self, dtype, op, tensors, equal_nan=True):
+        for t in tensors:
+            if dtype is torch.bfloat16:
+                a = t.cpu().to(torch.float32).numpy()
+            else:
+                a = t.cpu().numpy()
+
+            actual = op(t, t)
+            expected = op.ref(a, a)
+
+            # Crafts a custom error message for smaller, printable tensors
+            if t.numel() < 10:
+                msg = ("Failed to produce expected results! Input tensor was"
+                       " {0}, torch result is {1}, and reference result is"
+                       " {2}.").format(t, actual, expected)
+            else:
+                msg = None
+
+            exact_dtype = True
+            if not torch.can_cast(numpy_to_torch_dtype_dict[expected.dtype.type], dtype):
+                exact_dtype = False
+
+                if dtype in [torch.uint8, torch.int8, torch.bool]:
+                    # NOTE: For these dtypes, PyTorch computes in the default scalar type (float)
+                    # while NumPy computes in float16
+                    self.assertEqualHelper(actual, expected, msg, dtype=dtype,
+                                           exact_dtype=exact_dtype, rtol=1e-3, atol=1e-2)
+                    continue
+
+            self.assertEqualHelper(actual, expected, msg, dtype=dtype, equal_nan=equal_nan, exact_dtype=exact_dtype)
+
+    # Tests that the function and its (array-accepting) reference produce the same
+    #   values on a range of tensors, including empty tensors, scalar tensors,
+    #   1D tensors and a large 2D tensor with interesting and extremal values
+    #   and discontiguities.
+    @ops(binary_ufuncs)
+    def test_reference_numerics_normal(self, device, dtype, op):
```

Comment:
What if we start with just "test_reference_numerics_normal" and it simply calls make_tensor() to construct two large tensors of the same size for the moment?

Then we can think about how to extend the tests. I have some ideas that might simplify things.

This would let us remove the _hard and _extremal variants (for now) and cut the more complicated generator functions.